woocommerce/woocommerce-ios
# Enable/disable a product variation

Fixes #2511 

## Changes

- In the Networking layer, encoded `status` for `ProductVariation` 
- Added `UIImage.visibilityImage` using Gridicons' `visible` icon with a test case
- Added a case `status` to `ProductFormEditAction` for editing a variation's status
- In `ProductVariationFormActionsFactory`, always showed `status` action below the price settings
  - Updated unit tests to include this row
- In `ImageAndTitleAndTextTableViewCell`, added a different type of view model `SwitchableViewModel` to configure the new cell type with a switch toggle in the cell's accessory view
- In `DefaultProductFormTableViewModel`, configured the status row view model for a `ProductVariation`
- Added `status: ProductStatus` to `ProductFormDataModel` protocol, and return the value for a `Product`
- In `ProductFormTableViewDataSource`, added an action to handle status changes and configured the status row with `SwitchableViewModel`
- In `ProductFormTableViewModel`'s `ProductFormSection.SettingsRow`, added the corresponding `status` row that takes in a `SwitchableViewModel`
- In `ProductFormViewController`, configured the status action handling
- Added `func updateStatus(_ isEnabled: Bool)` to `ProductFormViewModelProtocol`, and implemented it for `ProductVariationFormViewModel` and no-op for `ProductFormViewModel`
  - Added unit tests for `ProductVariationFormViewModel`

## Testing

- Go to the Products tab
- Tap on a variable product with non-zero variations
- Tap on the variations row
- Tap on a variation that is disabled before in core --> the "Enable" row toggle should be off
- Turn on the toggle on the "Enabled" row
- Make sure there is a price for the variation
- Tap "Update" --> the variation status should be updated remotely ("Enabled" toggle in core)
- Navigate back to the variations list --> the variation cell should not show "Hidden" anymore (note that a price needs to be set, otherwise the variation is still hidden even though it's enabled)
- Tap on the same variation cell again --> the "Enable" row toggle should be on
- Turn off the toggle on the "Enabled" row
- Tap "Update" --> the variation status should be updated remotely ("Enabled" toggle in core)
- Navigate back to the variations list --> the variation cell should now show as "Hidden"

## Example screenshots

![simulator](https://user-images.githubusercontent.com/1945542/89244244-971d1680-d638-11ea-9d0c-d3631a2b0ad9.gif)

Mode | Disabled | Enabled
-- | -- | --
Dark | ![Simulator Screen Shot - iPhone 11 - 2020-08-04 at 09 53 13](https://user-images.githubusercontent.com/1945542/89244233-91273580-d638-11ea-8432-a710266a0e76.png) | ![Simulator Screen Shot - iPhone 11 - 2020-08-04 at 09 53 15](https://user-images.githubusercontent.com/1945542/89244238-95535300-d638-11ea-98e5-9f61a356b45a.png)
Light | ![Simulator Screen Shot - iPhone 11 - 2020-08-04 at 09 53 27](https://user-images.githubusercontent.com/1945542/89244242-95ebe980-d638-11ea-8cf6-b1d3c8433263.png) | ![Simulator Screen Shot - iPhone 11 - 2020-08-04 at 09 53 28](https://user-images.githubusercontent.com/1945542/89244243-96848000-d638-11ea-950a-ade7795e200e.png)



Update release notes:
